test/test_ncbi_taxa.py

```diff
@@ -16,304 +16,651 @@
             dict(
                 higher_nm="Canis vulpes", lower_nm="Vulpes vulpes schrenckii", tri=True
             ),
+            pytest.raises(ValueError),
+            None,
+            (
+                (
+                    ERROR,
+                    "Subspecies name (Vulpes vulpes schrenckii) appears to be trinomial"
+                    " but does not contain species name (Canis vulpes)",
+                ),
+            ),
+            id="subsp inconsistent species",
+        ),
+        pytest.param(
+            dict(higher_nm="Over long name", lower_nm="schrenckii", tri=True),
+            pytest.raises(ValueError),
+            None,
+            ((ERROR, "Species name (Over long name) appears to be too long"),),
+            id="subsp long species",
+        ),
+        pytest.param(
+            dict(higher_nm="Vulpes", lower_nm="Vulpes vulpes schrenckii", tri=True),
+            pytest.raises(ValueError),
             None,
+            ((ERROR, "Species name (Vulpes) too short"),),
+            id="subsp short species",
         ),
-        (dict(higher_nm="Over long name", lower_nm="schrenckii", tri=True), None),
-        (dict(higher_nm="Vulpes", lower_nm="Vulpes vulpes schrenckii", tri=True), None),
     ],
 )
-def test_construct_bi_or_tri(test_input, expected):
+def test_construct_bi_or_tri(caplog, test_input, raises, expected, expected_log):
     """Test function that constructs species binomials from species and genus names.
 
     We test that it can catch when the species name is already a binomial, and that it
     catches Candidatus names and handles them properly.
     """
+    with raises:
+        s_nm = taxa.construct_bi_or_tri(**test_input)
 
-    s_nm = taxa.construct_bi_or_tri(**test_input)
+    if isinstance(raises, does_not_raise):
+        assert s_nm == expected
 
-    assert s_nm == expected
+    log_check(caplog, expected_log)
 
 
-# Now test that the function logs errors correctly
+# ------------------------------------------
+# Testing taxon validators
+# ------------------------------------------
 @pytest.mark.parametrize(
-    "test_input,expected_log_entries",
-    [  # Fine so empty
-        (dict(higher_nm="Escherichia", lower_nm="coli", tri=False), ()),
-        (  # Over long name
-            dict(higher_nm="Over long genus name", lower_nm="vulpes", tri=False),
-            ((ERROR, "Genus name (Over long genus name) appears to be too long"),),
+    "tax_id,raises, leaf,n_tax",
+    [
+        pytest.param(
+            562,
+            does_not_raise(),
+            ("Escherichia coli", "species", 562, 561),
+            8,
+            id="Backbone leaf",
         ),
-        (  # Genus name not in binomial
-            dict(higher_nm="Canis", lower_nm="Vulpes vulpes", tri=False),
+        pytest.param(
+            562,
+            does_not_raise(),
+            ("Escherichia coli 1-110-08_S1_C1", "strain", 1444049, 562),
+            9,
+            id="Non backbone leaf",
```

Comment:
I might be misunderstanding the test but I don't get this one, I don't understand how searching `tax_id=562` will return `("Escherichia coli 1-110-08_S1_C1", "strain", 1444049, 562)` as the first element of the hierarchy?

Reply:
Test was borked and not actually testing that the `leaf` value matched. Which of course it didn't.